### DISCUSSION TITLE
Using Private Github Repo as NPM Dependency

### DISCUSSION THREAD

**Discussion by BrianHung** [vercel/community#143] [Help] [answered]
Is there a way of using a private github repo as an npm dependency?

Right now, deploying a project that has for example
```
"dependencies": {
  "xxx": "github:brianhung/xxx"
}
```
where `brianhung/xxx` is a private repo, results in a deployment error

```
04:06:42.091 | Exit code: 128
-- | --
04:06:42.091 | Command: git
04:06:42.091 | Arguments: ls-remote --tags --heads ssh://git@github.com/brianhung/xxx.git
04:06:42.091 | Directory: /vercel/workpath0
04:06:42.091 | Output:
04:06:42.091 | Host key verification failed.
04:06:42.091 | fatal: Could not read from remote repository.
04:06:42.091 | Please make sure you have the correct access rights
04:06:42.091 | and the repository exists.
```

I saw from this post https://github.com/vercel/vercel/issues/3728 that there's a way of adding _published_ private github and npm packages, but what about private github repos that aren't published at all?

**Comment by scucchiero** [comment]
I'm experiencing the same issue

**Comment by nlien** [comment]
Me too. Please educate us if there's a way to solve this. Otherwise it's a showstopper, unfortunately.

**Comment by planadecu** [comment]
Same issue here! It’s solved in bitbucket by allowing the platform to configire github credentials.

**Comment by wiiiimm** [comment]
@nlien @planadecu @scucchiero @BrianHung currently Vercel only supports private npm dependencies. Is there a reason why your private git repo cannot be published as a private package in npm?

**Reply by nlien** [reply]
No reason, except the extra steps required for 1) initial publish and 2) each code increment, which requires a new npm publish, as opposed to just pushing the code to GitHub.
Actually I took the steps to make a private npm package, but it would be nice if Vercel could support private repos as dependencies as well.

**Reply by BrianHung** [reply]
Same reasoning as nlien's: I split up a private project I'm deploying on Vercel into multiple private repos to make the codebase more modular and re-usable, and don't want to deal with the hassle of publishing those repos as packages every time I make an edit to the dependency repo.

**Comment by dgurns** [comment]
Ah I missed that somehow. It's an extra step from our side but it should work. Thanks. Perhaps it would be worth adding another reference or two in the docs to make that clear.

**Comment by planadecu** [comment]
Hi,

It costs an extra step in every dependency build process plus a paid
subscription per each NPM publisher.
For me it was easier to dirty copy paste the dependency in the main package
for now. Sad :(

El El jue, 3 dic 2020 a las 10:45, Dan Gurney <notifications@github.com>
escribió:

> Ah I missed that somehow. It's an extra step from our side but it should
> work. Thanks. Perhaps it would be worth adding another reference or two in
> the docs to make that clear.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/vercel/vercel/discussions/5274#discussioncomment-145770>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJIB5XCWPWCQZSZGQ66VHTSS5M4NANCNFSM4SHIPUJA>
> .
>
-- 


*---Jordi PlanadecursachTel: +34 616 85 46 60*

**Reply by nlien** [reply]
Yes, the subscription cost in a private npm repo is a really good reason for Vercel to also support private GitHub repos, IMO.

**Comment by mcsdevv** [comment]
Hello all, we will be looking to support the use of private dependencies from GitHub repositories - thanks for providing the reasons this would be of benefit to you.

It's not something that's likely to happen this year but when it arrives, we'll let you know on this thread.

**Comment by planadecu** [comment]
Happy to hear that! 👌

**Comment by Raaghu** [comment]


**Reply by fnavarrodev** [reply]
Hi, 
What's the content of config_env.js ?

**Reply by fnavarrodev** [reply]
I'm trying to execute a shell script like this:
```
#!/bin/bash
eval $(ssh-agent -s)
ssh-add <(echo "$GITHUB_TOKEN")
npm install
```
But I get this error:  sh: -c: line 0: syntax error near unexpected token `('

Using that shell script in other projects in Amplify that worked perfectly. Maybe this shell is different but I can't work it out. Anyone has a final solution?

**Comment by GerTea** [comment]
Any update on this?
@mcsdevv is there a timeline for this issue?

**Reply by mcsdevv** [reply]
No update on this just yet!

**Comment by CodingMeSwiftly** [comment]
![5h6k38](https://user-images.githubusercontent.com/6830248/126510383-3076daea-c961-4cf0-88d6-7d385f60f91c.jpg)

**Comment by barelyhuman** [comment]
There's a few things you'll have to do to get this setup as it's not the general course of action for vercel.

### Prerequisite
1. Generate a new SSH keypair specifically for this
2. Add the SSH keypair into your github account 

### Project Level Setup (Locally)
1. Write a script that can generate the ssh private key locally using environment variables
2. Next Steps on Vercel

### Vercel Portal
1. Setup the Project 
2. Settings => Environment Variables => Add your SSH Private Key here as a ENV Variable
3. In General Settings of the project, override the install step with running your script for generating the ssh key file first and then running yarn 
4. Yarn should be able to now use the provided key to fetch the deps. 

**Make sure you have the deps added in the following format**
```json
{
  "dependencies": {
    "private-dep": "git+ssh://git@github.com/barelyhuman/private-dep.git"
  }
}
```

To read the key during runtime , you can use the following or lookup more ways to inject a key during execution
```sh
ssh-agent bash -c 'ssh-add ./private-key; yarn'
```

The script for creating the private key 
```sh
#!/bin/bash

echo $GITHUB_SSH_KEY > private-key
```

**The above is totally theoretical and based on my knowledge of doing this with various CI/CD instances** 
It might not work but worth a try, let me know if you face any other issues during this. 

**Additional Notes**
You can also basically move the key generation and insertion into the same script and execute the script on the install command , but the separation is done so that I can help debug issues during build


**Edit: ** 
Forgot to add the bash script + additional notes

**Reply by mightymatth** [reply]
@barelyhuman 
> Settings => Environment Variables => Add your SSH Private Key here as a ENV Variable
```
Error:The total size of all Environment Variables (4.22KB) exceeds 4KB. Learn More →
```

I'm totally unable to build my application without some crazy workaround. 
In my case, I build my dependencies with npm where some of my dependencies have dependencies that are always set to be fetched via git protocol (not https), and we cannot override it unless changing `package-lock.json` after build. The way to fix it was to set a private ssh key of our git repository and everything worked fine.

**Reply by barelyhuman** [reply]
Oh yeah, the key would exceed on certain algos, should've put that in the message. 

Well the workarounds are needed since this scenario possible wasn't expected but I'm sure you can raise an issue of the same for allowing support private github packages somehow, they can then help you use github deploy keys to achieve something similar. 

Hopefully your solution works fine :)

**Comment by tirrth** [comment]
@mcsdevv Any update on this issue?

**Comment by tirrth** [comment]
Finally!! I was able to find a solution by following Netlify's documentation😄

**We just need to use the following format:**
```
{
     "dependencies": {
       "private-dep": "git+https://<github_personal_access_token>:x-oauth-basic@github.com/<user>/<repo>.git"
     }
}
```

Reference - https://docs.netlify.com/configure-builds/repo-permissions-linking/#:~:text=%23-,npm%20packages,-You%20can%20reference

**Reply by devon2018** [reply]
<img width="1034" alt="Screenshot 2021-12-18 at 22 46 57" src="https://user-images.githubusercontent.com/46196252/146657442-7a5e4ce7-9666-476d-ad43-368c1783517f.png">

Trying to use this solution, but seem to be getting a build error, works file locally though.

**Reply by sebastianrath** [reply]
Isn't the `package.json` often backed into build artifacts including the dependencies, e.g. Electron or vercel/pkg? I'm wondering if this could reveal access tokens by accident.

**Comment by afozbek** [comment]
Actually, all we have to do is create an .npmrc file on the root folder of your project.

Then update it based on your organization or account. 
```js
// .npmrc file

//npm.pkg.github.com/:_authToken=${GITHUB_TOKEN}
organization_name:registry=https://npm.pkg.github.com
```

This is for local, don't forget to add an environment variable to vercel as well

**Comment by jmurty** [comment]
Here's how I have been managing this successfully for the last several months, in case it helps others. It's basically what @afozbek suggests above, with a couple of tweaks.

In your local development environment, create a *.npmrc* file with contents like this, replacing `ORGANIZATION_NAME` and `GITHUB_TOKEN` with your own values:
```
; Use NPM registry by default
//registry.npmjs.org/

; GitHub Packages registry for your org's packages
ORGANIZATION_NAME:registry=https://npm.pkg.github.com/
//npm.pkg.github.com/:_authToken=GITHUB_TOKEN
```

For Vercel deployments, copy the contents of your *.npmrc* file to an `NPM_RC` environment variable in Vercel per the [Other Package Registries](https://vercel.com/support/articles/using-private-dependencies-with-vercel#other-package-registries) documentation.
The docs suggest using the command line `vc env add NPM_RC [environment] < ~/.npmrc` but I just copy/paste the contents of the *.npmrc* file into the environment variables web site UI, which also works fine.

**Comment by amyegan** [comment, marked as answer]
Hi everyone 👋 The original question is a couple of years old now, but some people have recently revived it. Since the original post, we have added a support article about [how to use private dependencies with Vercel](https://vercel.com/support/articles/using-private-dependencies-with-vercel). It covers git repos formatted as npm packages and private packages from other registries. 

Thank you to everyone who came here to discuss and share your techniques with the community!

Support article: https://vercel.com/support/articles/using-private-dependencies-with-vercel

**Reply by daviddelusenet** [reply]
@amyegan follow up on this, when using this syntax:

```
"package-name": "git+https://<token-username>:<token>@gitlab.com/<user>/<repo>.git"
```

How would you control which version is being installed?

**Reply by amyegan** [reply]
You could use something like `"npm": "git+https://github.com:npm/cli.git#v9.4.0"`. The[ npm-install doc](https://docs.npmjs.com/cli/v9/commands/npm-install) also has some details and examples the [`npm install <git remote url>`](https://docs.npmjs.com/cli/v9/commands/npm-install#:~:text=npm%20install%20%3Cgit%20remote%20url%3E) section. 

This Stack Overflow post about installing an [npm module from gitlab private repository](https://stackoverflow.com/questions/22988876/install-npm-module-from-gitlab-private-repository) and HackerNoon post about [Private NPM Packages](https://hackernoon.com/private-npm-packages-2e6579b37616) have some additional info that might be helpful.

**Reply by julesodea** [reply]
@amyegan, how would I update to the latest version without it being overwritten?

**Reply by amyegan** [reply]
Hi @julesodea! I'm not sure I understand what you're asking
